src/main/kotlin/org/rust/lang/core/types/ty/TyArray.kt

```diff
@@ -10,9 +10,10 @@ import org.rust.lang.core.resolve.ImplLookup
 import org.rust.lang.core.types.infer.TypeFolder
 import org.rust.lang.core.types.infer.TypeVisitor
 
-class TyArray(val base: Ty, val size: Int) : Ty {
+class TyArray(val base: Ty, val size: Long?) : Ty {
     override fun unifyWith(other: Ty, lookup: ImplLookup): UnifyResult =
-        if (other is TyArray && size == other.size) base.unifyWith(other.base, lookup) else UnifyResult.fail
+        if (other is TyArray && (size == null || other.size == null || size == other.size))
+            base.unifyWith(other.base, lookup) else UnifyResult.fail
```

Comment:
Usually when `if` is multiline we wrap it with `{ }`, but I want to merge it right now and `unifyWith` will be dropped soon :)